Lab | Inferential statistics - T-test & P-value
Instructions
One-tailed t-test - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on average than the machine currently used. To test that hypothesis, the times each machine takes to pack ten cartons are recorded. The results are in seconds in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t-test, does the data provide sufficient evidence to show if one machine is better than the other?

Matched Pairs Test - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment on your result.

Inferential statistics - ANOVA
Note: The following lab is divided into 2 sections which represent activities 3 and 4.

Part 1
In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

Context
Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder

State the null hypothesis
State the alternate hypothesis
What is the significance level
What are the degrees of freedom of the model, error terms, and total DoF
Data were collected randomly and provided to you in the table as shown: link to the image - Data

Part 2
In this section, use Python to conduct ANOVA.
What conclusions can you draw from the experiment and why?

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

In [13]:
machine = pd.read_csv('/Users/simonefischer/Ironhack/Week_7/Day_2/Afternoon/lab-t-tests-p-values/files_for_lab/machine.txt', sep='\t', encoding='utf-16')
machine.head()

New machine      Old machine
0         42.1             42.7
1         41.0             43.6
2         41.3             43.8
3         41.8             43.3
4         42.4             42.5

In [14]:
machine.shape

(10, 2)

In [18]:
machine.columns

Index(['New machine', '    Old machine'], dtype='object')

In [20]:
machine = machine.rename(columns={'    Old machine': 'old_machine', 'New machine': 'new_machine'})

In [21]:
machine

new_machine  old_machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [ ]:
##One-tailed t-test

##H0 there is no significant difference in performance before and after
##H1 there is significant difference in performance before and after


In [26]:
from scipy.stats import ttest_rel

t_statistic, p_value = ttest_rel(machine['new_machine'], machine['old_machine'], alternative = 'greater')

print("t-statistic:", t_statistic)
print("p-value:", p_value)

if p_value < 0.05:
    print("Reject the null hypothesis: There is a significant difference.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference.")


t-statistic: -3.0614273841115844
p-value: 0.9932298321741837
Fail to reject the null hypothesis: There is no significant difference.


In [24]:
pokemon = pd.read_csv('/Users/simonefischer/Ironhack/Week_7/Day_2/Afternoon/lab-t-tests-p-values/files_for_lab/pokemon.csv')

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [30]:
lower = ()

lower_columns = []

for column in pokemon.columns:
    lower_columns.append(column.lower())
pokemon.columns = lower_columns

display(pokemon.head())

#                   name type 1  type 2  total  hp  attack  defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   sp. atk  sp. def  speed  generation  legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [31]:
pokemon.shape

(800, 13)

In [ ]:
##Matched Pairs Test - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). 
##Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. 
##Our hypothesis is that the defense and attack scores are equal. 
##Compare the two columns to see if there is a statistically significant difference between them and comment on your result.

$$H0: \mu_{b} = \mu_{a} → \mu_{b} - \mu_{a} = 0$$
$$H1: \mu_{b} \neq \mu_{a} → \mu_{b} - \mu_{a} \neq 0$$

In [32]:
sample = pokemon.sample(30, random_state = 1)
sample['difference'] = sample['attack']-sample['defense']
sample.head()

#                       name  type 1  type 2  total  hp  attack  \
8      6  CharizardMega Charizard Y    Fire  Flying    634  78     104   
510  460                  Abomasnow   Grass     Ice    494  90      92   
175  161                    Sentret  Normal     NaN    215  35      46   
735  667                     Litleo    Fire  Normal    369  62      50   
242  224                  Octillery   Water     NaN    480  75     105   

     defense  sp. atk  sp. def  speed  generation  legendary  difference  
8         78      159      115    100           1      False          26  
510       75       92       85     60           4      False          17  
175       34       35       45     20           2      False          12  
735       58       73       54     72           6      False          -8  
242       75      105       75     45           2      False          30

In [33]:
sample_diff_mean, sample_diff_std = sample['difference'].mean(), sample['difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(2.8, 36.32079294288603)

In [34]:
t = sample_diff_mean / ( sample_diff_std / np.sqrt(sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 2.80
The standard deviation of our samples differences is: 36.32
Our t statistics is: 0.42


In [35]:
tc = st.t.ppf(1-(0.05/2),df= sample.shape[0] - 1)
tc

2.045229642132703

The statistic is 0.42 while the critical value is 2.04. Then, as 0.42 < 2.05 we accept the H0.

In [47]:
##Inferential statistics - ANOVA

#question: if changing the power of the plasma beam has any effect on the etching rate by the machine
#You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power.
#You can find the data anova_lab_data.xlsx file in the files_for_lab folder


data = pd.read_excel('/Users/simonefischer/Ironhack/Week_7/Day_2/Afternoon/lab-t-tests-p-values/files_for_lab/anova_lab_data.xlsx')

data.head()


Power   Etching Rate
0  160 W          5.43
1  180 W          6.24
2  200 W          8.79
3  160 W          5.71
4  180 W          6.71

In [41]:
data.shape

(15, 2)

In [58]:
data.columns

Index(['power ', 'etching_rate'], dtype='object')

In [48]:
lower = ()

lower_columns = []

for column in data.columns:
    lower_columns.append(column.lower())
data.columns = lower_columns

display(data.head())

power   etching rate
0  160 W          5.43
1  180 W          6.24
2  200 W          8.79
3  160 W          5.71
4  180 W          6.71

In [59]:
data = data.rename(columns={'etching rate': 'etching_rate', 'power ':'power'})


#State the null hypothesis

$$H0: \mu_{1}=\mu_{2}=\mu_{3}=⋯=\mu_{n}$$

--> all groups have the same means = there is no significant difference in the mean etching rate for different power levels of the plasma beam

#State the alternate hypothesis

$$H1: \mu_{i}\neq\mu_{j}$$ 

--> the means are significantly different = there is a significant difference in the mean etching rate for at least one power level of the plasma beam.

#What is the significance level

--> I will set it at 0.05

alpha = 0.05

#What are the degrees of freedom of the model, error terms, and total DoF


#Data were collected randomly and provided to you in the table as shown: link to the image - Data

#Part 2
#In this section, use Python to conduct ANOVA.
#What conclusions can you draw from the experiment and why?

In [60]:
data['power'].value_counts()

power
160 W    5
180 W    5
200 W    5
Name: count, dtype: int64

In [64]:
group_df = data.groupby('power')['etching_rate'].agg(etching_mean='mean',samples='size').reset_index()
group_df

power  etching_mean  samples
0  160 W         5.792        5
1  180 W         6.238        5
2  200 W         8.318        5

##Getting the test statistic (F-statistic) F = S2t / S2E 
​#Getting S2t
#The groups are power, and the "y" is the etching_rate

In [70]:
S2t = 0
for power in data['power'].unique():
    ng = len(data[data['power'] == power])
    S2t += ng * ((data[data['power'] == power]['etching_rate'].mean() - data['etching_rate'].mean()) ** 2)

S2t /= (data['power'].nunique() - 1)

print("The value of S2t is {:.2f}".format(S2t))

The value of S2t is 9.09


In [74]:
##Getting S2E

S2E = 0
for power in data['power'].unique():
    for rate in data[data['power'] == power]['etching_rate']:
        S2E += ( rate - data[data['power'] == power]['etching_rate'].mean() ) ** 2
S2E /= ( len(data) - data['power'].nunique() )

print()
print("The value of S2E is {:.2f}".format(S2E))


The value of S2E is 0.25


In [75]:
F = S2t / S2E
print("The value of F is {:.2f}".format(F))

The value of F is 36.88


In [65]:
##Degrees of freedom

d1 = data['power'].nunique() - 1
d2 = len(data) - data['power'].nunique()

print("Number of degrees of freedom d1: ",d1)
print("Number of degrees of freedom d2: ",d2)

Number of degrees of freedom d1:  2
Number of degrees of freedom d2:  12


In [76]:
st.f.cdf(F,dfn=d1, dfd=d2)

0.9999924934157276

Thus, the probability to get any value smaller or equal to F

$$P(x \le F=4.83)= 0.9999$$

In [85]:
##opposite:

1 - st.f.cdf(F,dfn=d1, dfd=d2)

##is 0

7.5065842723986975e-06

Therefore, the probability to get a value bigger than F is:

$$P(x > F) = 1 - P(x \le F) = 0.0 < 0.05$$

Therefore, we reject the H0

In [87]:
Fc = st.f.ppf(1-0.05,dfn=d1, dfd=d2)

print("The critical value which corresponds to an area of 0.05 is: {:.2f}".format(Fc))

The critical value which corresponds to an area of 0.05 is: 3.89


In [89]:
#F is bigger than the critical value, we reject H0.

In [ ]:
##Part 2: doing it with Python

In [90]:
from scipy.stats import f_oneway

In [96]:
group_df

power  etching_mean  samples
0  160 W         5.792        5
1  180 W         6.238        5
2  200 W         8.318        5

In [99]:
data['power'].unique()

array(['160 W', '180 W', '200 W'], dtype=object)

In [101]:
f_statistic, p_value = f_oneway(data[data['power'] == '160 W']['etching_rate'],data[data['power'] == '180 W']['etching_rate'],
                  data[data['power'] == '200 W']['etching_rate'])

In [102]:
f_statistic

36.87895470100505